### DISCUSSION TITLE
As the project grows larger and more routes and components are added, the speed of remix's rebuild process starts to noticeably slow down？

### DISCUSSION THREAD

**Discussion by yyong008** [remix-run/remix#6451] [General]
When working on a larger project, as the number of routes and components increases, along with the addition of third-party dependencies, the build time can increase to 20s+ or even longer. Making even a small change may require a rebuild of nearly 10+s. This significantly diminishes the development experience. Are there any solutions to this issue?

**Comment by lili21** [comment]
This issue really needs attention, and it is high priority.

**Reply by lili21** [reply]
maybe it is a foundation issue, is there any thing we can do to improve the esbuild rebuild performance? or maybe we should use another build tools?

**Reply by kiliman** [reply]
I think they need to add timing data to the build process. You can't improve something if you don't measure. It will be interesting to see where the bottlenecks are. Are you using any of the new CSS features?

**Reply by lili21** [reply]
I removed some CSS plugins from esbuild, but it didn't have a significant impact on the performance. To assess the rebuild performance, I ran esbuild without remix, but it wasn't as efficient as some other build tools. Even a simple project with antd dependency takes 700ms to rebuild

**Reply by lili21** [reply]
is esbuild the bottlenecks?

**Reply by lili21** [reply]
I created a reproduce repo, if anyone want to try it. https://github.com/lili21/esbuild-rebuild-perf

**Reply by pcattori** [reply]
Yea so in cases like this, we'd probably need to solve by doing pre-bundling for dependencies (fingerprinted by version) only once when that dependency is installed. There are trade-offs there (since treeshaking won't be as aggressive with pre-bundling), but that would make dev builds/rebuilds _much_ faster in the presence of large dependencies.

**Comment by pcattori** [comment]
The most helpful things for us to improve perf are reproductions that we can profile.

So far in 1.16.1, 1.17.0, and 1.17.1, we've added some performance improvements. We already have fixes and improvements for 1.18 as well. If anyone can reproduce performance issues on latest Remix version (or better yet, on `nightly`) and share a link, that would be awesome! 🙏

**Reply by yyong008** [reply]
I have a project and want to write the full stack with remix, but the packaging is getting slower and slower and I have to stop, this is not suitable, or I have to find another way to solve this problem, the project is in https://github.com/yyong008/remix-antd-admin

**Reply by pcattori** [reply]
For compilation to scale well with app size (i.e. `O(1)`) we'd need to do per-route compilation. It's definitely something we are looking into for v2.

I tried out your repo locally on my M1 Max MBP (check out that sweet new log output 🤩):

## remix build

<img width="751" alt="Screenshot 2023-06-16 at 3 11 19 PM" src="https://github.com/remix-run/remix/assets/1477317/537d396e-2395-46fd-a6bf-8a20d3dd1d9f">

## remix dev (`unstable_dev`)

<img width="930" alt="Screenshot 2023-06-16 at 3 11 05 PM" src="https://github.com/remix-run/remix/assets/1477317/c35c8e82-b996-43c1-a3f1-74ec8225af91">

**Reply by pcattori** [reply]
Did a bit more digging and added perf profiling:

<img width="751" alt="Screenshot 2023-06-16 at 3 15 30 PM" src="https://github.com/remix-run/remix/assets/1477317/3a04ba89-1839-442e-bf36-f158436f0e88">

Seems like CSS compilation accounts for 55% of your compilation time. ~2.7s with CSS, ~1.2s without CSS.

Here's with all CSS features disabled):

<img width="762" alt="Screenshot 2023-06-16 at 3 21 17 PM" src="https://github.com/remix-run/remix/assets/1477317/eab334e4-3edc-462b-9517-66fdeefaad7c">

**Reply by pcattori** [reply]
You are also using `antd`. Not sure how treeshakeable that library is. As mentioned here https://github.com/remix-run/remix/discussions/6539#discussioncomment-6194252 , you might be able to get better perf by avoiding named imports and deeply importing the specific things you use from `antd`.

**Reply by HosseinAgha** [reply]
@pcattori, how could you enable the debug logs?

**Comment by huseyn994** [comment]
I have same issue with latest remix version. Every single change takes more than 1minute for me. I use Windows.

**Reply by lili21** [reply]
are you using vite?

**Comment by Gsl-12** [comment]
This has become a major problem for us as well. We are a startup and have invested in the app enough that switching to next.Js would be next to impossible. 

This is urgent. Do we have any updates on this? 

For engineers and entrepreneurs considering remix, don't, for your sanity, until this is fixed.

**Reply by pcattori** [reply]
Have you tried using our [Vite plugin](https://remix.run/docs/en/main/future/vite)? It was specifically designed to address performance concerns raised in this discussion.

It's already [robust enough to use in production](https://twitter.com/kentcdodds/status/1747752059087327511), but still marked as `unstable` since some APIs are getting tweaked and renamed. We're actively working towards stabilizing it and are making good progress on that front.

**Reply by Gsl-12** [reply]
We are going to try this. Will update here

**Reply by pcattori** [reply]
If any performance issues remain after switching to Vite, try using the `--profile` flag as described in [_Vite > Performance_ docs](https://remix.run/docs/en/main/future/vite#performance).

**Reply by tombohub** [reply]
did you try vite?

**Reply by Gsl-12** [reply]
> did you try vite?

We did. To install the plug-in there was lots of work required to reach its requirements. It definitely helped. However we are moving away from Remix.

**Reply by RabeeAbuBaker** [reply]
> > did you try vite?
> 
> We did. To install the plug-in there was lots of work required to reach its requirements. It definitely helped. However we are moving away from Remix.

@Gsl-12 Performance with vite feels like magic to me, it drastically improved my experience. What is the reason for moving away even after trying vite? Just curious here...

**Reply by tombohub** [reply]
> > did you try vite?
> 
> We did. To install the plug-in there was lots of work required to reach its requirements. It definitely helped. However we are moving away from Remix.

Where are you going to?
